File: src/smartdesk-core/internal/domain/ticket.go
// Package domain holds the smartdesk-core domain model and pure business rules
// (state machine, SLA). It has no I/O dependencies so it is trivially testable.
//
// Shapes mirror openapi/core.yaml; field names are the wire contract.
package domain

import (
	"errors"
	"time"
)

// TicketStatus is the八态 lifecycle (core.yaml TicketStatus).
type TicketStatus string

const (
	StatusNew         TicketStatus = "new"
	StatusAccepted    TicketStatus = "accepted"
	StatusInProgress  TicketStatus = "in_progress"
	StatusPendingUser TicketStatus = "pending_user"
	StatusResolved    TicketStatus = "resolved"
	StatusClosed      TicketStatus = "closed"
	StatusSuspended   TicketStatus = "suspended"
	StatusCancelled   TicketStatus = "cancelled"
)

// Terminal reports whether the status is a终态 (closed/cancelled, OQ-8).
func (s TicketStatus) Terminal() bool {
	return s == StatusClosed || s == StatusCancelled
}

// Priority levels (core.yaml Priority).
type Priority string

const (
	P1 Priority = "P1"
	P2 Priority = "P2"
	P3 Priority = "P3"
	P4 Priority = "P4"
)

// Valid reports whether p is a known priority.
func (p Priority) Valid() bool {
	switch p {
	case P1, P2, P3, P4:
		return true
	}
	return false
}

// ReopenWindow is the window after closing during which a requester may reopen
// (OQ-13: 7 days).
const ReopenWindow = 7 * 24 * time.Hour

// Domain-level sentinel errors mapped to HTTP status by the api layer.
var (
	ErrNotFound          = errors.New("not found")
	ErrIllegalTransition = errors.New("illegal transition")
	ErrUnprocessable     = errors.New("unprocessable")
	ErrForbidden         = errors.New("forbidden")
)

// Ticket is the authoritative工单 record (core.yaml Ticket).
type Ticket struct {
	ID          string       `json:"id"`
	OrgID       string       `json:"org_id"`
	Number      string       `json:"number"`
	Title       string       `json:"title"`
	Description string       `json:"description,omitempty"`
	RequesterID string       `json:"requester_id,omitempty"`
	AssigneeID  *string      `json:"assignee_id"`
	GroupID     *string      `json:"group_id"`
	CategoryID  *string      `json:"category_id"`
	Priority    Priority     `json:"priority"`
	Status      TicketStatus `json:"status"`
	Source      string       `json:"source"`
	ReopenCount int          `json:"reopen_count"`
	CsatScore   *int         `json:"csat_score"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`

	// ClosedAt is internal bookkeeping for the reopen window; not serialized.
	ClosedAt *time.Time `json:"-"`
}

// Comment is a public reply or internal note (core.yaml Comment).
type Comment struct {
	ID         string    `json:"id"`
	TicketID   string    `json:"ticket_id"`
	AuthorID   string    `json:"author_id"`
	Body       string    `json:"body"`
	Visibility string    `json:"visibility"` // public | internal
	Mentions   []string  `json:"mentions,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
}

// Assignment records a分派/改派/转派/升级 (core.yaml Assignment).
type Assignment struct {
	ID        string    `json:"id"`
	TicketID  string    `json:"ticket_id"`
	Kind      string    `json:"kind"` // manual | auto | reassign | escalate
	ToUserID  *string   `json:"to_user_id"`
	ToGroupID *string   `json:"to_group_id"`
	Reason    string    `json:"reason,omitempty"`
	ActorID   string    `json:"actor_id"`
	CreatedAt time.Time `json:"created_at"`
}

// TimelineEntry is an append-only audit record (core.yaml TimelineEntry, §3).
type TimelineEntry struct {
	ID        string         `json:"id"`
	TicketID  string         `json:"ticket_id"`
	EventType string         `json:"event_type"`
	ActorID   *string        `json:"actor_id"`
	Payload   map[string]any `json:"payload"`
	CreatedAt time.Time      `json:"created_at"`
}

// Category is a taxonomy node (core.yaml Category).
type Category struct {
	ID       string  `json:"id"`
	ParentID *string `json:"parent_id"`
	Code     string  `json:"code,omitempty"`
	Name     string  `json:"name"`
	Active   bool    `json:"active"`
	Sort     int     `json:"sort"`
}

// User is RBAC master data (core.yaml User).
type User struct {
	ID          string   `json:"id"`
	Username    string   `json:"username"`
	Email       string   `json:"email,omitempty"`
	DisplayName string   `json:"display_name"`
	Status      string   `json:"status"` // active | disabled
	Roles       []string `json:"roles"`
}

File: src/smartdesk-web/src/types/ticket.ts
/**
 * 工单相关类型定义（生产代码使用）
 *
 * 从 e2e/fixtures/types 提取核心类型，供生产代码引用
 * 避免生产代码直接依赖 e2e fixtures
 */

/** 工单状态 */
export type TicketStatus =
  | 'new'
  | 'accepted'
  | 'in_progress'
  | 'pending_user'
  | 'resolved'
  | 'closed'
  | 'suspended'
  | 'cancelled';

/** 工单优先级 */
export type Priority = 'P1' | 'P2' | 'P3' | 'P4';

/** 用户角色 */
export type UserRole = 'requester' | 'agent' | 'lead' | 'manager' | 'admin';

/** 评论可见性 */
export type CommentVisibility = 'public' | 'internal';

/** 工单模型 */
export interface Ticket {
  id: string;
  number?: string;
  title: string;
  status: TicketStatus;
  priority: Priority;
  assignee_id: string | null;
  category_id: string | null;
  created_at: string;
  org_id?: string;
  requester_id?: string;
  description?: string;
  updated_at?: string;
  resolved_at?: string | null;
  closed_at?: string | null;
  first_response_at?: string | null;
  tags?: string[];
  suggestion?: any | null;
}

/** 工单聚合（详情） */
export interface TicketAggregate extends Ticket {
  requester?: {
    id: string;
    display_name: string;
    email?: string | null;
  };
  requester_name?: string;
  comments_count?: number;
}

/** 工单列表响应 */
export interface TicketPage {
  items: Ticket[];
  total: number;
  page: number;
  page_size: number;
}

/** 评论模型 */
export interface Comment {
  id: string;
  body: string;
  visibility: CommentVisibility;
  author_id: string;
  created_at: string;
  author_name?: string;
  ticket_id?: string;
  updated_at?: string;
}

/** 评论列表响应 */
export interface CommentPage {
  items: Comment[];
  total: number;
  page: number;
  page_size: number;
}

/** 当前用户信息 */
export interface Me {
  user_id: string;
  username: string;
  display_name: string;
  roles: UserRole[];
}
